java/memory/src/main/java/org/apache/arrow/memory/util/ArrowBufPointer.java

```diff
@@ -0,0 +1,132 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.arrow.memory.util;
+
+import io.netty.buffer.ArrowBuf;
+
+/**
+ * Pointer to a memory region within an {@link io.netty.buffer.ArrowBuf}.
+ * It will be used as the basis for calculating hash code within a vector, and equality determination.
+ */
+public final class ArrowBufPointer {
```

Comment:
Did you consider adding hashcode and equality to ArrowBuf directly?

Reply:
@emkornfield good question.
Adding hash code & equality to ArrowBuf is also a good choice.

I think there are several reasons for this data structure:
1. We want to compute and compare an arbitrary sub-area of the ArrowBuf, not the complete buffer. 
2. The algorithm to compute the hash code should be configurable to be suitable for different scenarios.
3. We need a way to show that the data area is invalid.
4. ArrowBuf is the key data structure, so we do not want to add overhead (like new members) to it.

Reply:
- 1 and 4 can be solved by using a slice of an arrow buf (address/length adjusted to the data element).
- 3 can be solved by using a null value (same as for ArrowBufPointer)

The problem with slices is that there is perf overhead due to the refCnt incr/decr. So, I'm fine with the ArrowBufPointer approach.

Reply:
OK, so is this approach dangerous then in the sense that we could have a dangling pointer?

Reply:
@emkornfield I think you are right.
It is possible to have dangling pointers (as in C++). The users can check it by examining the reference count of the underlying ArrowBuf. 

However, for most scenarios, I think the users have sufficient knowledge about the underlying ArrowBuf, so the checks can be avoided.